File: JAVA/10_HeapCreation&HM/Lecture45/HashMap.java
package Lecture45;

import java.util.*;

public class HashMap {
    // assuming key is int, value is string

    class HMNode{
        Integer key;
        String value;
        HMNode(int key, String value){
            this.key = key;
            this.value = value;
        }
    }

    private LinkedList<HMNode>[] buckets;

    private int size;

    private void initbucket(int nob){
        size = 0;
        buckets = new LinkedList[nob];

        for(int i = 0; i < buckets.length; i++){
            buckets[i] = new LinkedList<>();
        }
    }

    private int hashfn(Integer key){
        int hc = key;
        int bi = Math.abs(hc) % buckets.length;
        return bi;
    }

    private int findWithinbucket(int bi, Integer key){
        for(int di = 0; di < buckets[bi].size(); di++){
            HMNode node = buckets[bi].get(di);
            if(node.key.equals(key)){
                return di;
            }
        }

        return -1;
    }

    private void rehash(){
        int onob = buckets.length;
        LinkedList<HMNode>[] obuckets = buckets;

        initbucket(2 * onob);

        for(int bi = 0; bi < onob; bi++){
            for(int di = 0; di < obuckets[bi].size(); di++){
                HMNode node=obuckets[bi].get(di);
                put(node.key, node.value);
            }
        }
    }
    HashMap(){
        initbucket(4);
        size = 0;
    }

    public void put(Integer key, String value){
        int bi = hashfn(key);
        int di = findWithinbucket(bi, key);
        if(di == -1){
            HMNode node = new HMNode(key, value);
            buckets[bi].addLast(node);
            size++;
            double lambda = size * 1.0 / buckets.length;
            if(lambda > 2.0){
                rehash();
            }
        } 
        else{
            HMNode node = buckets[bi].get(di);
            node.value = value;
        }
    }

    public String get(Integer key){
        int bi = hashfn(key);
        int di = findWithinbucket(bi, key);
        if(di == -1){
            return null;
        }
        else{
            HMNode node =  buckets[bi].get(di);
            return node.value;
        }
    }

    public String remove(Integer key){
        int bi = hashfn(key);
        int di = findWithinbucket(bi, key);
        if(di == -1){
            return null;
        }
        else{
            String value = buckets[bi].get(di).value;
            buckets[bi].remove(di);
            size--;
            return value;
        }
    }

    public boolean containsKey(Integer key){
        int bi = hashfn(key);
        int di = findWithinbucket(bi, key);
        if(di == -1){
            return false;
        }
        else{
            return true;
        }
    }

    public ArrayList<Integer> keyset(){
        ArrayList<Integer> keys = new ArrayList<>();
        for(int bi = 0; bi < buckets.length; bi++){
            for(int di = 0; di < buckets[bi].size(); di++){
                keys.add(buckets[bi].get(di).key);
            }
        }
        return keys;
    }

    public void display(){
        System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~");
        for(int bi = 0; bi < buckets.length; bi++){
            System.out.print("b" + bi + " -> ");
            for(int di = 0; di < buckets[bi].size(); di++){
                HMNode node = buckets[bi].get(di);
                System.out.print("[" + node.key + ", " + node.value + "], ");
            }
            System.out.println(".");
        }
        System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~");
    }



    public static void main(String[] args){
        HashMap map = new HashMap();
        map.put(1, "a");
        map.put(2, "b");
        map.put(3, "c");
        map.put(4, "d");
        map.put(5, "e");
        map.put(6, "f");
        map.put(7, "g");
        map.put(8, "h");
        map.display();

        map.put(9, "i");
        map.display();
    }
    
}

File: C++/DS2/Lec45/HashMap.cpp
#include<iostream>
#include<vector>
#include<string>
#include<list>

using namespace std;

class HMNode
{
    public:
        int key;
        string value;
        HMNode(int key, string value)
        {
            this->key = key;
            this->value = value;
        }
};

class hashMap
{
    private:
        list<HMNode>* buckets = NULL;
        int size;
        int nob;
        int hashfn(int key)
        {
            int hc = key;
            int bi = abs(hc) % nob;
            return bi;
        }

        list<HMNode>::iterator findWithinbucket(int bi, int key)
        {
            int di = 0;
            for(auto itr = buckets[bi].begin(); itr != buckets[bi].end(); itr++)
            {
                HMNode node = *itr;
                if(node.key == key)
                {
                    return itr;
                }
            }
            return buckets[bi].end();
        }

        void rehash(){
            int onob = nob;
            list<HMNode>* obuckets = buckets;

            nob = 2 * onob;
            buckets = new list<HMNode>[nob];
            size = 0;

            for(int bi = 0; bi < onob; bi++)
            {
                for(auto di = obuckets[bi].begin(); di != obuckets[bi].end(); di++)
                {
                    (*this)[di->key] = di->value;
                }
            }
            delete[] obuckets;
        }
    public:

        hashMap()
        {
            nob = 4;
            buckets = new list<HMNode>[nob];
            size = 0;
        }

        // for get and put 
        string& operator[](int key)
        {
            int bi = hashfn(key);
            auto itr = findWithinbucket(bi, key);
            if(itr == buckets[bi].end())
            {
                HMNode node(key, "");
                buckets[bi].push_back(node);
                size++;

                double lambda = (size * 1.0 )/ nob;
                if(lambda > 2.0)
                {
                    rehash();
                }

                bi = hashfn(key);
                itr = findWithinbucket(bi, key);
                return itr->value;
            }
            else
            {
                return itr->value;
            }
        }

        int count(int key)
        {
            int bi = hashfn(key);
            auto itr = findWithinbucket(bi, key);
            if(itr == buckets[bi].end())
            {
                return 0;
            }
            else
            {
                return 1;
            }
        }

        void erase(int key)
        {
            int bi = hashfn(key);
            auto itr = findWithinbucket(bi, key);
            if(itr == buckets[bi].end()){
                return;
            }
            else
            {
                size--;
                buckets[bi].erase(itr);
            }
        }

        vector<int> keyset(){
            vector<int> keys;
            for(int bi = 0; bi < nob; bi++){
                for(auto itr = buckets[bi].begin(); itr != buckets[bi].end(); itr++)
                {
                    keys.push_back(itr->key);
                }
            }
            return keys;
        }

        void display()
        {
            cout << "~~~~~~~~~~~~~~~~~~~~~~~~~~" << endl;
            for(int bi = 0; bi < nob; bi++)
            {
                cout << "b" << bi << " -> ";
                for(auto itr = buckets[bi].begin(); itr != buckets[bi].end(); itr++)
                {
                    cout << "[" << itr->key << ", " << itr->value << "]" << " ,";
                }
                cout << "." << endl;
            }
            cout << "~~~~~~~~~~~~~~~~~~~~~~~~~~" << endl;
        }

};


int main(int argv, char** argc)
{
    hashMap map;
    map[1] = "A";
    map[2] = "B";
    map[3] = "C";
    map[4] = "D";
    map[5] = "E";
    map[6] = "F";
    map[7] = "G";
    map[8] = "H";
    map.display();
    map[9] = "I";
    map.display();

}